interfaces/ingress/v0/interface_tests/provider_tests.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+from scenario import State, Relation
+
+from interface_tester.interface_test import interface_test_case, SchemaConfig
+
+
+@interface_test_case(
+    event='ingress-relation-created',
+    role='provider',
+    schema=SchemaConfig.empty
+)
+def test_no_data_on_created(output_state: State):
+    # nothing happens on created: databags are empty
+    return
+
+
+@interface_test_case(
+    event='ingress-relation-joined',
+    role='provider',
+    schema=SchemaConfig.empty
+)
+def test_no_data_on_joined(output_state: State):
+    # nothing happens on joined: databags are empty
+    return
+
+
+@interface_test_case(
+    event='ingress-relation-changed',
+    role='provider',
+    input_state=State(
+        relations=[Relation(
+            # todo: endpoint is unknown/overwritten: find an elegant way to omit it here.
+            #  perhaps input state is too general: we only need this relation's db contents:
+            endpoint='ingress',
+            interface='ingress',
+            remote_app_name='remote',
+            remote_app_data={
+                'host': '0.0.0.42',
+                'model': 'bar',
+                'name': 'remote/0',
+                'port': '42',
+            }
+        )]
+    )
+)
+def test_data_published_on_changed_remote_valid(output_state: State):
+    return  # schema validation is enough for now
+
+
+@interface_test_case(
+    event='ingress-relation-changed',
+    role='provider',
+    input_state=State(relations=[Relation(
+        endpoint='ingress',
+        interface='ingress',
+        remote_app_name='remote',
+        remote_app_data={
+            'port': '42',
+            'bubble': 'rubble'
+        }
+    )]
+    ),
+    schema=SchemaConfig.empty
+)
+def test_data_published_on_changed_remote_invalid(output_state: State):
+    # on changed, if the remote side has sent INvalid data: local side didn't publish anything either.
+    return
```

Comment:
Is there a test in here that validates anything?

Reply:
yes. 
see README_INTERFACE_TESTS:
```
## Steps of an interface test
An interface test such as the one above has three steps:

- verify that the scenario runs
- verify that the output state (`state_out`) is valid (by the test-writer's definition)
- validate against the relation schemas

Failure in any of the three steps means a failed test.
```

The decorated function represents step 2: it's a validator that can contain any custom state validation logic. In this case we happen not to need any.
Suppose that it was part of the ingress:provider specification that in that situation the charm needs to set blocked status, the tests could have been:
```
def test_data_published_on_changed_remote_invalid(output_state: State):
    assert isinstance(output_state.status.unit, BlockedStatus)
```

Reply:
this should be more clear in the readme now.